test/box-luatest/gh_8650_add_is_sync_option_test.lua

```diff
@@ -0,0 +1,319 @@
+        end)
+    end)
+
+    -- Sync transaction
+    t.assert_error_msg_content_equals('Quorum collection for a synchronous ' ..
+                                      'transaction is timed out',
+    function()
+        cg.master:exec(function()
+            box.atomic({is_sync = true}, function()
+                box.space.test:insert{10, 'is_sync = true'}
+            end)
+        end)
+    end)
+
+    t.assert_equals(cg.master:exec(function()
+        return box.space.test:select()
+    end), {{1, 'is_sync = false'}, {2, 'is_sync = false'},
+           {3, 'is_sync = true'}, {4, 'is_sync = true'},
+           {9, "is_sync = false"}})
+end
+
+local function inject_iproto_begin_or_commit(c, request, is_sync)
+    local header = msgpack.encode({
+        [box.iproto.key.REQUEST_TYPE] = box.iproto.type[request],
+        [box.iproto.key.SYNC] = c:_next_sync(),
+        [box.iproto.key.STREAM_ID] = c._stream_id or 0,
+    })
+    local body = msgpack.encode({
+        [box.iproto.key.IS_SYNC] = is_sync,
+    })
+    local size = msgpack.encode(#header + #body)
+    local request = size .. header .. body
+    return c:_inject(request)
+end
+
+g.test_iproto_is_sync = function(cg)
+    local c = net.connect(cg.master.net_box_uri)
+    local s = c:new_stream()
+
+    cg.master:exec(function()
+        box.schema.space.create('test')
+        box.space.test:create_index('pk')
+        box.ctl.promote()
+    end)
+
+    -- Throwing iproto errors if is_sync is set incorrectly
+    local err_msg = "Illegal parameters, is_sync can only be true"
+    t.assert_error_msg_content_equals(err_msg, function()
+        inject_iproto_begin_or_commit(s, 'BEGIN', false)
+    end)
+
+    err_msg = "Invalid MsgPack - request body"
+    t.assert_error_msg_content_equals(err_msg, function()
+        inject_iproto_begin_or_commit(s, 'BEGIN', 'foo')
+    end)
+
+    err_msg = "Illegal parameters, is_sync can only be true"
+    t.assert_error_msg_content_equals(err_msg, function()
+        inject_iproto_begin_or_commit(s, 'COMMIT', false)
+    end)
+
+    err_msg = "Invalid MsgPack - request body"
+    t.assert_error_msg_content_equals(err_msg, function()
+        inject_iproto_begin_or_commit(s, 'COMMIT', 'foo')
+    end)
+
+    -- is_sync is set correctly
+    s:begin()
+    s.space.test:insert({1, 'is_sync = false'})
+    s:commit()
+
+    s:begin({is_sync = true})
+    s.space.test:insert({2, 'is_sync = true'})
+    s:commit()
+
+    s:begin()
+    s.space.test:insert({3, 'is_sync = true'})
+    s:commit({is_sync = true})
+
+    cg.replica:stop()
+
+    -- Async transaction
+    s:begin()
+    s.space.test:insert({4, 'is_sync = false'})
+    s:commit()
+
+    -- Sync transaction
+    t.assert_error_msg_content_equals('Quorum collection for a synchronous ' ..
+                                      'transaction is timed out',
+    function()
+        s:begin({is_sync = true})
+        s.space.test:insert({5, 'is_sync = true'})
+        s:commit()
+        s:begin()
```

Comment:
Same as above, the first begin-commit always fails due to a timeout, the second begin-commit is not tested.

Reply:
Fixed